README
ts2mpa
======
By Nicholas Humfrey <njh@aelius.com>

ts2mpa is a simple tool to extract MPEG Audio from a MPEG-2 Transport Stream.


Usage: ts2mpa [options] <infile> <outfile>
    -h             Help - this message.
    -q             Quiet - don't print messages to stderr.
    -p <pid>       Choose a specific transport stream PID.
    -s <streamid>  Choose a specific PES stream ID.



Examples
--------

The following examples are based on DVB-T in London, tuned to 
BBC Radio coming from the the Crystal Palace transmitter:

Record BBC Radio 4 to disk:
dvbstream -o -f 529833330 439 | ts2mpa - recording.mp2

Play BBC Radio 1 through soundcard:
dvbstream -o -f 529833330 436 | ts2mpa -q - - | mpg123 -




Todo
----

- Add support for resyncing transport stream
- Automatic output file name choosing
- Progress bar?



ChangeLog
Version 0.2 (2008-01-10)
------------------------

- Added -q option to disable messages
- Added -s option to choose a stream id
- Added -p option to choose a PID
- Fix for Linux, but adding #include <signal.h>
- Added support for STDIN and STDOUT


Version 0.1 (2008-01-10)
------------------------

- Initial Release



ts2mpa.c
/* 

	ts2mpa.c
	(C) Nicholas J Humfrey <njh@aelius.com> 2008
	
	Copyright notice:
	
	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.
	
	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
    
*/


#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>

#include "ts2mpa.h"
#include "mpa_header.h"

int Quiet = 0;
int Interrupted = 0;



// Check to see if a PES header looks like a valid MPEG Audio one
static int validate_pes_header( int pid, unsigned char* buf_ptr, int buf_len )
{
	unsigned char stream_id = 0x00;

	// Does it have the right magic?
	if( PES_PACKET_SYNC_BYTE1(buf_ptr) != 0x00 ||
	    PES_PACKET_SYNC_BYTE2(buf_ptr) != 0x00 ||
	    PES_PACKET_SYNC_BYTE3(buf_ptr) != 0x01 )
	{
		if (!Quiet) fprintf(stderr, "ts2mpa: Invalid PES header (pid: %d).  \n", pid);
		return 0;
	}
	
	// Is it MPEG Audio?
	stream_id = PES_PACKET_STREAM_ID(buf_ptr);
	if (stream_id < 0xC0 || stream_id > 0xDF) {
		if (!Quiet) fprintf(stderr, "ts2mpa: Ignoring non-mpegaudio stream (pid: %d, stream id: 0x%x).\n", pid, stream_id);
		return 0;
	}

	// Check PES Extension header 
	if( PES_PACKET_SYNC_CODE(buf_ptr) != 0x2 )
	{
		if (!Quiet) fprintf(stderr, "ts2mpa: Invalid sync code PES extension header (pid: %d, stream id: 0x%x).\n", pid, stream_id);
		return 0;
	}

	// Reject scrambled packets
	if( PES_PACKET_SCRAMBLED(buf_ptr) )
	{
		if (!Quiet) fprintf(stderr, "ts2mpa: PES payload is scrambled (pid: %d, stream id: 0x%x).\n", pid, stream_id);
		return 0;
	}

	// It is valid
	return 1;
}


// Extract the PES payload and send it to the output file
static void extract_pes_payload( ts2mpa_t *ts2mpa, unsigned char *pes_ptr, size_t pes_len, int start_of_pes ) 
{
	unsigned char* es_ptr=NULL;
	size_t es_len=0;
	
	
	// Start of a PES header?
	if ( start_of_pes ) {
		unsigned int pes_total_len = PES_PACKET_LEN(pes_ptr);
		size_t pes_header_len = PES_PACKET_HEAD_LEN(pes_ptr);
		unsigned char stream_id = PES_PACKET_STREAM_ID(pes_ptr);
	
		// Check that it has a valid header
		if (!validate_pes_header( ts2mpa->pid, pes_ptr, pes_len )) return;
		
		// Stream IDs in range 0xC0-0xDF are MPEG audio
		if( stream_id != ts2mpa->pes_stream_id )
		{	
			if (ts2mpa->pes_stream_id == -1) {
				// keep the first stream we see
				ts2mpa->pes_stream_id = stream_id;	
				if (!Quiet)
					fprintf(stderr, "ts2mpa: Found valid PES audio packet (pid: %d, stream id: 0x%x, length: %u)\n",
									ts2mpa->pid, stream_id, pes_total_len);
			} else {
				if (!Quiet)
					fprintf(stderr, "ts2mpa: Ignoring additional audio stream ID 0x%x (pid: %d).\n",
							stream_id, ts2mpa->pid);
				return;
			}
		}
	
		// Store the length of the PES packet payload
		ts2mpa->pes_remaining = pes_total_len - (2+pes_header_len);
	
		// Keep pointer to ES data in this packet
		es_ptr = pes_ptr+(9+pes_header_len);
		es_len = pes_len-(9+pes_header_len);
	

	} else if (ts2mpa->pes_stream_id) {
	
		// Only output data once we have seen a PES header
		es_ptr = pes_ptr;
		es_len = pes_len;
	
		// Are we are the end of the PES packet?
		if (es_len>ts2mpa->pes_remaining) {
			es_len=ts2mpa->pes_remaining;
		}
		
	}

	
	// Got some data to write out?
	if (es_ptr) {
		
		// Subtract the amount remaining in current PES packet
		ts2mpa->pes_remaining -= es_len;
	
		// Scan through Elementary Stream (ES) 
		// and try and find MPEG audio stream header
		while (!ts2mpa->synced && es_len>=4) {
		
			// Valid header?
			if (mpa_header_parse(es_ptr, &ts2mpa->mpah)) {

				// Looks good, we have gained sync.
				if (!Quiet && ts2mpa->never_synced) {
					fprintf(stderr, "ts2mpa: ");
					mpa_header_print( &ts2mpa->mpah );
					fprintf(stderr, "ts2mpa: MPEG Audio Framesize: %d bytes\n", ts2mpa->mpah.framesize);
				}
				ts2mpa->synced = 1;
				ts2mpa->never_synced = 0;

			} else {
				// Skip byte
				es_len--;
				es_ptr++;
			}
		}
		
		
		// If stream is synced then put data info buffer
		if (ts2mpa->synced) {
			int written = 0;
			
			// Write out the data
			written = fwrite( es_ptr, es_len, 1, ts2mpa->output );
			if (written<=0) {
				perror("Error: failed to write stream out");
				exit(-2);
			}
			ts2mpa->total_bytes += written;
		}
	}

}



static void ts_continuity_check( ts2mpa_t *ts2mpa, int ts_cc ) 
{
	if (ts2mpa->continuity_count != ts_cc) {
	
		// Only display an error after we gain sync
		if (ts2mpa->synced) {
			if (!Quiet) fprintf(stderr, "ts2mpa: Warning, TS continuity error (pid: %d)    \n", ts2mpa->pid );
			ts2mpa->synced=0;
		}
		ts2mpa->continuity_count = ts_cc;
	}

	ts2mpa->continuity_count++;
	if (ts2mpa->continuity_count==16)
		ts2mpa->continuity_count=0;
}



static void process_ts_packets( ts2mpa_t *ts2mpa )
{
	unsigned char buf[TS_PACKET_SIZE];
	unsigned char* pes_ptr=NULL;
	size_t pes_len;
	size_t count;

	while ( !Interrupted ) {
		
		count = fread(buf, TS_PACKET_SIZE, 1, ts2mpa->input);
		if (count==0) break;
		
		// Display the number of packets processed
		if (!Quiet) fprintf(stderr, "ts2mpa: packets processed: %lu\r", ts2mpa->total_packets);
		ts2mpa->total_packets++;
		
		// Check the sync-byte
		if (TS_PACKET_SYNC_BYTE(buf) != 0x47) {
			fprintf(stderr,"ts2mpa: Lost Transport Stream syncronisation - aborting.    \n");
			// FIXME: try and regain synchronisation
			break;
		}
		
		// Transport error?
		if ( TS_PACKET_TRANS_ERROR(buf) ) {
			if (!Quiet) fprintf(stderr, "ts2mpa: Warning, transport error in PID %d.    \n", TS_PACKET_PID(buf));
			ts2mpa->synced = 0;
			continue;
		}			

		// Scrambled?
		if ( TS_PACKET_SCRAMBLING(buf) ) {
			if (!Quiet) fprintf(stderr, "ts2mpa: Warning, PID %d is scrambled.    \n", TS_PACKET_PID(buf));
			continue;
		}	

		// Location of and size of PES payload
		pes_ptr = &buf[4];
		pes_len = TS_PACKET_SIZE - 4;

		// Check for adaptation field?
		if (TS_PACKET_ADAPTATION(buf)==0x1) {
			// Payload only, no adaptation field
		} else if (TS_PACKET_ADAPTATION(buf)==0x2) {
			// Adaptation field only, no payload
			continue;
		} else if (TS_PACKET_ADAPTATION(buf)==0x3) {
			// Adaptation field AND payload
			pes_ptr += (TS_PACKET_ADAPT_LEN(buf) + 1);
			pes_len -= (TS_PACKET_ADAPT_LEN(buf) + 1);
		}


		// Check we know about the payload
		if (TS_PACKET_PID(buf) == 0x1FFF) {
			// Ignore NULL package
			continue;
		}
		
		// No chosen PID yet?
		if (ts2mpa->pid == -1 && TS_PACKET_PAYLOAD_START(buf)) {

			// Does this one look good ?
			if (TS_PACKET_PAYLOAD_START(buf) && 
			    validate_pes_header( TS_PACKET_PID(buf), pes_ptr, pes_len ))
			{
				// Looks good, use this one
				ts2mpa->pid = TS_PACKET_PID(buf);
			}
		}

		// Process the packet, if it is the PID we are interested in		
		if (TS_PACKET_PID(buf) == ts2mpa->pid) {

			// Continuity check
			ts_continuity_check( ts2mpa, TS_PACKET_CONT_COUNT(buf) );
		
			// Extract PES payload and write it to output
			extract_pes_payload( ts2mpa, pes_ptr, pes_len, TS_PACKET_PAYLOAD_START(buf) );
		}

	}

	
}

static ts2mpa_t * init_ts2mpa_t()
{
	ts2mpa_t *ts2mpa = malloc( sizeof(ts2mpa_t) );
	if (ts2mpa==NULL) {
		perror("Failed to allocate memory for ts2mpa_t");
		exit(-3);
	}
	
	// Zero the memory
	bzero( ts2mpa, sizeof(ts2mpa_t) );
	
	// Initialise defaults
	ts2mpa->input = NULL;
	ts2mpa->output = NULL;
	ts2mpa->pid = -1;
	ts2mpa->synced = 0;
	ts2mpa->never_synced = 1;
	ts2mpa->continuity_count = -1;
	ts2mpa->pes_stream_id = -1;
	ts2mpa->pes_remaining = 0;
	ts2mpa->total_bytes = 0;
	ts2mpa->total_packets = 0;

	return ts2mpa;
}


static void usage()
{
	fprintf( stderr, "Usage: ts2mpa [options] <infile> <outfile>\n" );
	fprintf( stderr, "    -h             Help - this message.\n" );
	fprintf( stderr, "    -q             Quiet - don't print messages to stderr.\n" );
	fprintf( stderr, "    -p <pid>       Choose a specific transport stream PID.\n" );
	fprintf( stderr, "    -s <streamid>  Choose a specific PES stream ID.\n" );
	exit(-1);
}


static int parse_value( char* str )
{
	int value = 0;
	
	// Is it a hexadecimal?
	if (str[0] == '0' && (str[1] == 'x' || str[1] == 'X')) {
		str += 2;
		if (sscanf(str, "%x", &value)<=0) {
			fprintf(stderr, "ts2mpa: failed to parse hexadeicmal number.\n");
			usage();
		} else {
			return value;
		}
	} else {
		return atoi( str );
	}
	
	
	// Shouldn't get here
	return -1;
}


static void parse_cmd_line( ts2mpa_t *ts2mpa, int argc, char** argv )
{
	int ch;


	// Parse the options/switches
	while ((ch = getopt(argc, argv, "p:s:qh?")) != -1)
	switch (ch) {
		case 'q':
			Quiet = 1;
		break;
	
		case 'p':
			ts2mpa->pid = parse_value( optarg );
			if (ts2mpa->pid <= 0) {
				fprintf(stderr, "ts2mpa: Invalid Transport Stream PID: %s\n", optarg);
				exit(-1);
			}
		break;

		case 's':
			ts2mpa->pes_stream_id = parse_value( optarg );
			if (ts2mpa->pes_stream_id <= 0) {
				fprintf(stderr, "ts2mpa: Invalid PES Stream ID: %s\n", optarg);
				exit(-1);
			}
		break;

		case '?':
		case 'h':
		default:
			usage();
	}


	// Open the input file
	if (argc-optind < 1) {
		fprintf(stderr, "ts2mpa: missing input and output files.\n");
		usage();
	} else if ( strncmp( argv[optind], "-", 1 ) == 0 ) {
		// Use STDIN
		ts2mpa->input = stdin;
	} else {
		ts2mpa->input = fopen( argv[optind], "rb" );
		if (ts2mpa->input==NULL) {
			perror("ts2mpa: Failed to open input file");
			exit(-2);
		}
	}

	// Open the output file
	if (argc-optind < 2) {
		fprintf(stderr, "ts2mpa: missing output file.\n");
		usage();
	} else if ( strncmp( argv[optind+1], "-", 1 ) == 0 ) {
		// Use STDOUT
		ts2mpa->output = stdout;
	} else {
		ts2mpa->output = fopen( argv[optind+1], "wb" );
		if (ts2mpa->output==NULL) {
			perror("ts2mpa: Failed to open output file");
			exit(-2);
		}
	}
}

static void termination_handler(int signum)
{
	if (signum==SIGINT) fprintf(stderr, "ts2mpa: Recieved SIGINT, aborting.\n");
	else if (signum==SIGTERM) fprintf(stderr, "ts2mpa: Recieved SIGTERM, aborting.\n");
	else if (signum==SIGHUP) fprintf(stderr, "ts2mpa: Recieved SIGHUP, aborting.\n");
	
	Interrupted = 1;
}


int main( int argc, char** argv )
{
	ts2mpa_t *ts2mpa = init_ts2mpa_t();

	// Parse the command-line parameters
	parse_cmd_line( ts2mpa, argc, argv );
	
	// Setup signal handling - so we exit cleanly
	if (signal (SIGINT, termination_handler) == SIG_IGN)
		signal (SIGINT, SIG_IGN);
	if (signal (SIGHUP, termination_handler) == SIG_IGN)
		signal (SIGHUP, SIG_IGN);
	if (signal (SIGTERM, termination_handler) == SIG_IGN)
		signal (SIGTERM, SIG_IGN);

	// Hard work happens here
	process_ts_packets( ts2mpa );
	
	if (!Quiet) fprintf(stderr, "\nts2mpa: Total written: %lu bytes\n", ts2mpa->total_bytes);
	
	// Close the input and output files
	fclose( ts2mpa->input );
	fclose( ts2mpa->output );
	
	free(ts2mpa);
	
	// Success
	return 0;
}



ts2mpa.h
/* 

	ts2mpa.h
	(C) Nicholas J Humfrey <njh@aelius.com> 2008
	
	Copyright notice:
	
	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.
	
	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.
	
	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
    
*/

#ifndef _TS2MPA_H
#define _TS2MPA_H

#include "mpa_header.h"



typedef struct ts2mpa_s {
	
	FILE* input;
	FILE* output;
	
	int pid;
	int synced;
	int never_synced;
	int continuity_count;
	int pes_stream_id;
	int pes_remaining;
	unsigned long total_bytes;
	unsigned long total_packets;
	
	mpa_header_t mpah;
	
} ts2mpa_t;



// The size of MPEG2 TS packets
#define TS_PACKET_SIZE			188

/*
	Macros for accessing MPEG-2 TS packet headers
*/
#define TS_PACKET_SYNC_BYTE(b)		(b[0])
#define TS_PACKET_TRANS_ERROR(b)	((b[1]&0x80)>>7)
#define TS_PACKET_PAYLOAD_START(b)	((b[1]&0x40)>>6)
#define TS_PACKET_PRIORITY(b)		((b[1]&0x20)>>4)
#define TS_PACKET_PID(b)			(((b[1]&0x1F)<<8) | b[2])

#define TS_PACKET_SCRAMBLING(b)		((b[3]&0xC0)>>6)
#define TS_PACKET_ADAPTATION(b)		((b[3]&0x30)>>4)
#define TS_PACKET_CONT_COUNT(b)		((b[3]&0x0F)>>0)
#define TS_PACKET_ADAPT_LEN(b)		(b[4])



/*
	Macros for accessing MPEG-2 PES packet headers
*/
#define PES_PACKET_SYNC_BYTE1(b)	(b[0])
#define PES_PACKET_SYNC_BYTE2(b)	(b[1])
#define PES_PACKET_SYNC_BYTE3(b)	(b[2])
#define PES_PACKET_STREAM_ID(b)		(b[3])
#define PES_PACKET_LEN(b)			((b[4] << 8) | b[5])

#define PES_PACKET_SYNC_CODE(b)		((b[6] & 0xC0) >> 6)
#define PES_PACKET_SCRAMBLED(b)		((b[6] & 0x30) >> 4)
#define PES_PACKET_PRIORITY(b)		((b[6] & 0x08) >> 3)
#define PES_PACKET_ALIGNMENT(b)		((b[6] & 0x04) >> 2)
#define PES_PACKET_COPYRIGHT(b)		((b[6] & 0x02) >> 1)
#define PES_PACKET_ORIGINAL(b)		((b[6] & 0x01) >> 0)

#define PES_PACKET_PTS_DTS(b)		((b[7] & 0xC0) >> 6)
#define PES_PACKET_ESCR(b)			((b[7] & 0x20) >> 5)
#define PES_PACKET_ESR(b)			((b[7] & 0x10) >> 4)
#define PES_PACKET_DSM_TRICK(b)		((b[7] & 0x8) >> 3)
#define PES_PACKET_ADD_COPY(b)		((b[7] & 0x4) >> 2)
#define PES_PACKET_CRC(b)			((b[7] & 0x2) >> 1)
#define PES_PACKET_EXTEN(b)			((b[7] & 0x1) >> 0)
#define PES_PACKET_HEAD_LEN(b)		(b[8])

#define PES_PACKET_PTS(b)		((uint32_t)((b[9] & 0x0E) << 29) | \
					 (uint32_t)(b[10] << 22) | \
					 (uint32_t)((b[11] & 0xFE) << 14) | \
					 (uint32_t)(b[12] << 7) | \
					 (uint32_t)(b[13] >> 1))

#define PES_PACKET_DTS(b)		((uint32_t)((b[14] & 0x0E) << 29) | \
					 (uint32_t)(b[15] << 22) | \
					 (uint32_t)((b[16] & 0xFE) << 14) | \
					 (uint32_t)(b[17] << 7) | \
					 (uint32_t)(b[18] >> 1))



#endif



mpa_header.h
/*
 *  MPEG Audio Header Parser
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *  
 */

#ifndef _MPA_HEADER_H
#define _MPA_HEADER_H



typedef struct {
	unsigned int syncword;
	unsigned int layer;
	unsigned int version;
	unsigned int error_protection;
	unsigned int bitrate_index;
	unsigned int samplerate_index;
	unsigned int padding;
	unsigned int extension;
	unsigned int mode;
	unsigned int mode_ext;
	unsigned int copyright;
	unsigned int original;
	unsigned int emphasis;
	unsigned int channels;
	unsigned int bitrate;
	unsigned int samplerate;
	unsigned int samples;
	unsigned int framesize;
} mpa_header_t;



// Get parse the header of a frame of mpeg audio
int mpa_header_parse( const unsigned char* buf, mpa_header_t *mh);

void mpa_header_print( mpa_header_t *mh );
void mpa_header_debug( mpa_header_t *mh );



#endif


mpa_header.c
/*
 *  MPEG Audio Header Parser
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *  
 */


/*
    MPEG Audio frame handling courtesy of Scott Manley
    Borrowed from libshout 2.1 / mp3.c 
    With a few changes and fixes
 */
 
 
#include <stdio.h>
#include <stdlib.h>

#include "mpa_header.h"


#define MPA_MODE_STEREO		0
#define MPA_MODE_JOINT		1
#define MPA_MODE_DUAL		2
#define MPA_MODE_MONO		3


static const unsigned int bitrate[3][3][16] =
{
	{ // MPEG 1
		{0, 32, 64, 96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448, 0}, // Layer 1
		{0, 32, 48, 56,  64,  80,  96, 112, 128, 160, 192, 224, 256, 320, 384, 0}, // Layer 2
		{0, 32, 40, 48,  56,  64,  80,  96, 112, 128, 160, 192, 224, 256, 320, 0}  // Layer 3
	},
	{ // MPEG 2
		{0, 32, 48, 56,  64,  80,  96, 112, 128, 144, 160, 176, 192, 224, 256, 0}, // Layer 1
		{0,  8, 16, 24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0}, // Layer 2
		{0,  8, 16, 24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0}  // Layer 3
	},
	{ // MPEG 2.5
		{0, 32, 48, 56,  64,  80,  96, 112, 128, 144, 160, 176, 192, 224, 256, 0}, // Layer 1
		{0,  8, 16, 24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0}, // Layer 2
		{0,  8, 16, 24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0}  // Layer 3
	}
};

static const unsigned int samplerate[3][4] =
{
	{ 44100, 48000, 32000, 0 }, // MPEG 1
	{ 22050, 24000, 16000, 0 }, // MPEG 2
	{ 11025, 12000,  8000, 0 }  // MPEG 2.5
};


static void parse_header(mpa_header_t *mh, unsigned int header)
{
	mh->syncword = (header >> 20) & 0x0fff;

	mh->version = 2 - ((header >> 19) & 0x01);
	if ((mh->syncword & 0x01) == 0)
		mh->version = 3;

	mh->layer = 4-((header >> 17) & 0x03);
	if (mh->layer==4)
		mh->layer=0;
		
	mh->error_protection = ((header >> 16) & 0x01) ? 0 : 1;
	mh->bitrate_index = (header >> 12) & 0x0F;
	mh->samplerate_index = (header >> 10) & 0x03;
	mh->padding = (header >> 9) & 0x01;
	mh->extension = (header >> 8) & 0x01;
	mh->mode = (header >> 6) & 0x03;
	mh->mode_ext = (header >> 4) & 0x03;
	mh->copyright = (header >> 3) & 0x01;
	mh->original = (header >> 2) & 0x01;
	mh->emphasis = header & 0x03;

	if (mh->layer && mh->version) {
		mh->bitrate = bitrate[mh->version-1][mh->layer-1][mh->bitrate_index];
		mh->samplerate = samplerate[mh->version-1][mh->samplerate_index];
	} else {
		mh->bitrate = 0;
		mh->samplerate = 0;
	}

	if (mh->mode == MPA_MODE_MONO)
		mh->channels = 1;
	else
		mh->channels = 2;

	if (mh->version == 1)
		mh->samples = 1152;
	else
		mh->samples = 576;

	if(mh->samplerate)
		mh->framesize = (mh->samples * mh->bitrate * 1000 / mh->samplerate) / 8 + mh->padding;
}


// print out all field values for debuging
void mpa_header_debug( mpa_header_t *mh )
{
	if (mh->version==1)			fprintf(stderr, "  version=MPEG-1\n");
	else if (mh->version==2)	fprintf(stderr, "  version=MPEG-2\n");
	else if (mh->version==3)	fprintf(stderr, "  version=MPEG-2.5\n");
	else 						fprintf(stderr, "  version=unknown\n");

	fprintf(stderr, "  layer=%d\n", mh->layer);
	
	if (mh->mode==MPA_MODE_STEREO)		fprintf(stderr, "  mode=Stereo\n");
	else if (mh->mode==MPA_MODE_JOINT)	fprintf(stderr, "  mode=Joint Stereo\n");
	else if (mh->mode==MPA_MODE_DUAL)	fprintf(stderr, "  mode=Dual\n");
	else if (mh->mode==MPA_MODE_MONO)	fprintf(stderr, "  mode=Mono\n");
	else 								fprintf(stderr, "  mode=unknown\n");

	fprintf(stderr, "  error_protection=%d\n", mh->error_protection);
	fprintf(stderr, "  padding=%d\n", mh->padding);
	fprintf(stderr, "  extension=%d\n", mh->extension);
	fprintf(stderr, "  mode_ext=%d\n", mh->mode_ext);
	fprintf(stderr, "  copyright=%d\n", mh->copyright);
	fprintf(stderr, "  original=%d\n", mh->original);
	fprintf(stderr, "  channels=%d\n", mh->channels);
	fprintf(stderr, "  bitrate=%d\n", mh->bitrate);
	fprintf(stderr, "  samplerate=%d\n", mh->samplerate);
	fprintf(stderr, "  samples=%d\n", mh->samples);
	fprintf(stderr, "  framesize=%d\n", mh->framesize);
}

// concise informational string
void mpa_header_print( mpa_header_t *mh )
{
	if (mh->version==1)			fprintf(stderr, "MPEG-1 ");
	else if (mh->version==2)	fprintf(stderr, "MPEG-2 ");
	else if (mh->version==3)	fprintf(stderr, "MPEG-2.5 ");
	else 						fprintf(stderr, "MPEG-?? ");

	fprintf(stderr, "layer %d, ", mh->layer);
	fprintf(stderr, "%d kbps, ", mh->bitrate);
	fprintf(stderr, "%d Hz, ", mh->samplerate);
	
	if (mh->mode==MPA_MODE_STEREO)		fprintf(stderr, "Stereo");
	else if (mh->mode==MPA_MODE_JOINT)	fprintf(stderr, "Joint Stereo");
	else if (mh->mode==MPA_MODE_DUAL)	fprintf(stderr, "Dual");
	else if (mh->mode==MPA_MODE_MONO)	fprintf(stderr, "Mono");

	fprintf(stderr, "\n");
}


// Parse mpeg audio header
// returns 1 if valid, or 0 if invalid
int mpa_header_parse( const unsigned char* buf, mpa_header_t *mh)
{
	unsigned long head;
		
	/* Quick check */
	if (buf[0] != 0xFF)
		return 0;

	/* Put the first four bytes into an integer */
	head = ((unsigned int)buf[0] << 24) | 
		   ((unsigned int)buf[1] << 16) |
		   ((unsigned int)buf[2] << 8)  |
		   ((unsigned int)buf[3]);


	/* fill out the header struct */
	parse_header(mh, head);

	/* check for syncword */
	if ((mh->syncword & 0x0ffe) != 0x0ffe)
		return 0;

	/* make sure layer is sane */
	if (mh->layer == 0)
		return 0;

	/* make sure version is sane */
	if (mh->version == 0)
		return 0;

	/* make sure bitrate is sane */
	if (mh->bitrate == 0)
		return 0;

	/* make sure samplerate is sane */
	if (mh->samplerate == 0)
		return 0;

	return 1;
}


